### PR TITLE
Enable image resources from App_Resources in iOS

### PR DESCRIPTION
**Pull request by Fatme**
Fixes https://github.com/NativeScript/nativescript-cli/issues/520
https://github.com/NativeScript/ios-runtime/pull/202

### CODE REVIEW

**Comment by ErjanGavalji** [comment]
:+1:

File: lib/services/ios-project-service.ts [hunk 1]
```diff
@@ -221,11 +219,58 @@ class IOSProjectService implements  IPlatformProjectService {
...
+			this.$logger.trace("Current images from App_Resources");
+			this.$logger.trace(currentImageNames);
+
+			if(imageNames.length !== currentImageNames.length) {
+				let imagesToAdd = _.difference(currentImageNames, imageNames);
+				this.$logger.trace(`New images to add into xcode project: ${ imagesToAdd.length !== 0 ? imagesToAdd.join(", ") : "" }`);			
+				_.each(imagesToAdd, image => project.addResourceFile(path.join(this.platformData.appResourcesDestinationDirectoryPath, image)));
+				
+				let imagesToRemove = _.difference(imageNames, currentImageNames);
+				this.$logger.trace(`Images to remove from xcode project: ${ imagesToRemove.length !== 0 ? imagesToRemove.join(", ") : "" }`);
```
Comment (teobugslayer) [inline lib/services/ios-project-service.ts]: You can omit the length check.

```
builder@ab-linuxbpc-1:~/.ssh$ node
> [].join(", ")
''
>
```

File: lib/services/ios-project-service.ts [hunk 2]
```diff
@@ -221,11 +219,58 @@ class IOSProjectService implements  IPlatformProjectService {
...
+		return (() => {
+			let project = this.xcodeproj;
+			let resources = project.pbxGroupByName("Resources");
+			let references = project.pbxFileReferenceSection();
+			
+			let images = _(<any[]>resources.children)
+				.map(resource => references[resource.value])
+				.value();
+			
+			let imageNames = _.map(images, image => image.name.replace(/\"/g, ""));
```
Comment (teobugslayer) [inline lib/services/ios-project-service.ts]: You use images only here, so you can join the two maps together

Comment (teobugslayer) [inline lib/services/ios-project-service.ts]: I think this variable is poorly named. the log.trace line bellow suggests a better name: xcodeProjectImages

File: lib/services/ios-project-service.ts [hunk 3]
```diff
@@ -221,11 +219,58 @@ class IOSProjectService implements  IPlatformProjectService {
...
+			let currentImageNames = this.$fs.readDirectory(this.platformData.appResourcesDestinationDirectoryPath).wait();
```
Comment (teobugslayer) [inline lib/services/ios-project-service.ts]: Again, the name should be something like "appResourcesImages"

File: lib/services/ios-project-service.ts [hunk 4]
```diff
@@ -221,11 +219,58 @@ class IOSProjectService implements  IPlatformProjectService {
...
+			
+			this.$logger.trace("Images from xcode project");
+			this.$logger.trace(imageNames);
+			this.$logger.trace("Current images from App_Resources");
+			this.$logger.trace(currentImageNames);
+
+			if(imageNames.length !== currentImageNames.length) {
```
Comment (teobugslayer) [inline lib/services/ios-project-service.ts]: What happens if I delete a file in App_Resources and add another one? Or rather, if I rename an existing file?

File: lib/services/ios-project-service.ts (hunk 2 above)
Comment (rosen-vladimirov) [inline lib/services/ios-project-service.ts]: Or maybe:
```TypeScript
let imageNames = _.map(<any[]>resources.children, resource => references[resource.value].name.replace(/\"/g, ""));
```

File: lib/services/ios-project-service.ts (hunk 4 above)
Comment (rosen-vladimirov) [inline lib/services/ios-project-service.ts]: is it correct to have this check? What if the count is the same, but some of the names are different?

File: lib/services/android-project-service.ts [hunk 5]
```diff
@@ -273,6 +273,10 @@ class AndroidProjectService implements IPlatformProjectService {
 	public getFrameworkFilesExtensions(): string[] {
 		return [".jar", ".dat"];
 	}
+	
+	public prepareProject(): IFuture<void> {
+		return (() => { }).future<void>()();
```
Comment (ligaz) [inline lib/services/android-project-service.ts]: Maybe `_.noop()` here.

File: lib/services/ios-project-service.ts [hunk 6]
```diff
@@ -182,13 +182,11 @@ class IOSProjectService implements  IPlatformProjectService {
             this.$fs.ensureDirectoryExists(targetPath).wait();
             shell.cp("-R", libraryPath, targetPath);
 
-            var pbxProjPath = path.join(platformData.projectRoot, this.$projectData.projectName + ".xcodeproj", "project.pbxproj");
-            var project = new xcode.project(pbxProjPath);
-            project.parseSync();
+           let project = this.xcodeproj;
```
Comment (ligaz) [inline lib/services/ios-project-service.ts]: Indent me.

File: lib/services/ios-project-service.ts [hunk 7]
```diff
@@ -221,11 +219,58 @@ class IOSProjectService implements  IPlatformProjectService {
...
 		}).future<void>()();
 	}
+	
+	public prepareProject(): IFuture<void> {
+		return (() => {
+			let project = this.xcodeproj;
+			let resources = project.pbxGroupByName("Resources");
+			let references = project.pbxFileReferenceSection();
+			
+			let images = _(<any[]>resources.children)
```
Comment (ligaz) [inline lib/services/ios-project-service.ts]: Why complicating this with lodash when map is built-in?

File: lib/services/ios-project-service.ts [hunk 8]
```diff
@@ -221,11 +219,58 @@ class IOSProjectService implements  IPlatformProjectService {
...
+				.map(resource => references[resource.value])
+				.value();
+			
+			let imageNames = _.map(images, image => image.name.replace(/\"/g, ""));
+			let currentImageNames = this.$fs.readDirectory(this.platformData.appResourcesDestinationDirectoryPath).wait();
+			
+			this.$logger.trace("Images from xcode project");
```
Comment (ligaz) [inline lib/services/ios-project-service.ts]: Xcode is the correct spelling.

File: lib/services/android-project-service.ts (hunk 5 above)
Comment (rosen-vladimirov) [inline lib/services/android-project-service.ts]: we have to return future, maybe ```Future.fromResult()``` :)

File: lib/services/ios-project-service.ts (hunk 4 above)
Comment (ligaz) [inline lib/services/ios-project-service.ts]: Maybe `_.difference`

File: lib/services/ios-project-service.ts [hunk 9]
```diff
@@ -221,11 +219,58 @@ class IOSProjectService implements  IPlatformProjectService {
...
+	}
+	
+	private get xcodeproj(): any {
+		let project = new xcode.project(this.pbxProjPath);
+		project.parseSync();
+		
+		return project;
+	}
+	
+	private saveXcodeproj(project: any): IFuture<void> {
```
Comment (ligaz) [inline lib/services/ios-project-service.ts]: This is better named `savePbxProj`. The same goes for the getter.

File: lib/services/ios-project-service.ts [hunk 10]
```diff
@@ -221,11 +219,55 @@ class IOSProjectService implements  IPlatformProjectService {
...
+			
+		}).future<void>()();
+	}
+	
+	private get pbxProjPath(): string {
+		return path.join(this.platformData.projectRoot, this.$projectData.projectName + ".xcodeproj", "project.pbxproj");
+	}
+	
+	private get pbxProj(): any {
+		let project = new xcode.project(this.pbxProjPath);
```
Comment (rosen-vladimirov) [inline lib/services/ios-project-service.ts]: is it expected to return new project every time you try using the property pbxProj?

Comment (Fatme) [inline lib/services/ios-project-service.ts]: This function is called only one time - https://github.com/NativeScript/nativescript-cli/pull/590/files#diff-b673f137deb4afccccc0bd5f1572c2a0R229

Comment (ligaz) [inline lib/services/ios-project-service.ts]: So we can make it a function instead of property then.

**Comment by teobugslayer** [comment]
:+1: after addressing @rosen-vladimirov and @ligaz comments.

**Comment by rosen-vladimirov** [comment]
:+1:

**Pull request merged**